### PR TITLE
Add initial AVX2 support.

### PR DESCRIPTION
**Pull request by mtl1979**
* Add detection for AVX2
* Use AVX2 optimized slide_hash when available

### CODE REVIEW

File: win32/Makefile.msc [hunk 1]
```diff
@@ -23,7 +23,7 @@ AR = lib
 RC = rc
 CP = copy /y
 CFLAGS  = -nologo -MD -W3 -O2 -Oy- -Zi -Fd"zlib" $(LOC)
-WFLAGS  = -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -DX86_PCLMULQDQ_CRC -DX86_SSE2 -DX86_CPUID -DX86_SSE42_CRC_HASH -DUNALIGNED_OK -DX86_QUICK_STRATEGY
+WFLAGS  = -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -DX86_PCLMULQDQ_CRC -DX86_SSE2 -DX86_CPUID -DX86_SSE42_CRC_INTRIN -DX86_SSE42_CRC_HASH -Dx86_AVX2 -DUNALIGNED_OK -DX86_QUICK_STRATEGY
```
Comment (nmoinvaz) [inline win32/Makefile.msc]: I think this should be -DX86_AVX2 (uppercase X)

Comment (mtl1979) [reply win32/Makefile.msc]: Fixed.

**Comment by Dead2** [comment]
Since slide_avx.c is not your code, I'd appreciate it if that is added first as a separate commit with a short comment of where it came from.
I'll let the unrelated change in Makefile.msc slide ;)

**Comment by mtl1979** [comment]
slide_avx.c is modified from slide_sse.c ... Only thing I changed was the intrinsic names and increments in the loops...

**Comment by Dead2** [comment]
Ah, that explains it.

I think you should add rewrite the copyright text at the top then. Not sure how much you changed, but it sounds like it is no longer Intel's work. Alternatively you could just add a comment something like "Based on Intels slide_sse implementation, modified to use AVX2 instead".
Currently it reads like it is Intels work and that you modified it a little to fit zlib-ng datatypes and such.

**Comment by mtl1979** [comment]
I tried to be as safe as possible regarding copyright laws... US laws are stricter than EU laws. There is certain percentage that needs to be different so that copyright text can be truncated and in my opinion there is not enough difference to warrant truncating the original copyright.

**Comment by Dead2** [comment]
You can still add a comment like the one I suggested though.

**Comment by mtl1979** [comment]
@Dead2 Better?

**Comment by Dead2** [comment]
Perfect :)

**Comment by mtl1979** [comment]
I can't run benchmarks as my Linux doesn't have AVX2, only Windows machine has...

**Comment by Dead2** [comment]
I'll benchmark it soonTM :)

**Comment by Dead2** [comment]
In my benchmarks this only improves performance on average by ~0.55% with cmake while performance seems unchanged with configure (Might be caused by configure building it static).

While it improves performance on average, I notice that it is not consistently better. I run each compression level 11 times, and some levels on average end up faster while some are slower.

So, do you want to have another look at things to see whether anything can be improved first?

**Comment by mtl1979** [comment]
On my Windows machine, AVX2 code is marginally faster than SSE2 equivalent code when CPU load is 100%. I suspect the speed improvement is less if CPU load is lower than 100% due to most modern CPUs reducing CPU clock speed if CPU load is less than 100%.

**Comment by nmoinvaz** [comment]
@Dead2 what are you using for benchmarking?

**Comment by Dead2** [comment]
I use my own python tool that also disables power saving features such as c-states beyond C1 and forces a fixed clock speed to avoid turbo and/or throttling from influencing things. The core is also disabled as far as the linux scheduler is concerned, so no other tasks (or interrputs) interrupts the benchmark (with the exception of some RCU stuff I guess). And the files are located on a tmpfs (ramdisk), so repeatability is pretty good with mostly sub 0.1% variation between runs.

For timings I run minigzip through perf and parse its output with only the cputime (not time elapsed) counter enabled.

Mostly what seems to cause some variation in benchmark runs is cache (other things on the system might eat up a lot of L3 cache for a short time), possibly memory alignment, and syscalls. Therefore I filter out the X worst results for each level. Currently I run 11 tests per level and filter out the 5 worst.

To avoid fork() and such taking up a lot of the time, I currently use a 40MiB file for testing.
Today I have been running every PR with both cmake and configure, each taking ~5-6min including the perf-testing.

**Comment by mtl1979** [comment]
I would use multiple test files of different size when testing AVX2 as it heavily depends on file size due to short time spending inside slide_*.

**Comment by Dead2** [comment]
@mtl1979 Does that mean you are done and this is ready for merging?

**Comment by mtl1979** [comment]
@Dead2 If I do something, it makes more sense to do separate pull request...

**Comment by gdevenyi** [comment]
>  I currently use a 40MiB file for testing.

Please take a look at https://github.com/zlib-ng/zlib-ng/issues/326#issuecomment-578414906 for some corner-case file types which might be worth adding to a benchmarking suite.

**Comment by Dead2** [comment]
@gdevenyi My 40MiB file is actually a tar file containing multiple file types (testing small files separately leads to inaccuracy due to fork() and syscalls having a much bigger impact), but adding some new files to one of my testfiles would be very much interesting.  :)

**Comment by gdevenyi** [comment]
@Dead2 ah cheers, that makes some sense. In the linked case the important thing is the "padded with zeros" files, which we expect act pretty differently than regular files, and hence are a corner case in benchmarking.

**Comment by Dead2** [comment]
For some reason, codecov only reports a 4% coverage on the codechanges from this patch.

**Comment by Dead2** [comment]
Have a look here: https://codecov.io/gh/zlib-ng/zlib-ng/commit/53f9257a24362386e75504836eaeb2788360b1e8

It looks like cpuid gets run, but the check in `fill_window_sse` is always false so `slide_hash_sse2` gets used instead.

**Comment by nmoinvaz** [comment]
@Dead2 are you saying we need a CI configuration with SSE2 turned off?

**Comment by Dead2** [comment]
No, but that is a good point, we probably need to be able to disable AVX2 so SSE2 can be tested..

But currently it seems like AVX2 detection fails in the CI setup, it would be very weird if github actions didn't support AVX2 since that was introduced with Haswell years ago and I don't think there would be any reason to disable it.. Strange to say the least.

**Comment by mtl1979** [comment]
cpuid fails on WSL too. /proc/cpuinfo was fixed in WSL some time ago.

**Comment by nmoinvaz** [comment]
I am also seeing AVX2 disabled in GitHub Actions. 

When doing CPU ID check for 7, it returns all zeros. 

```
4: cpu 0 - ebx 756e6547, ecx 6c65746e, edx 49656e69 - max 13
4: cpu 1 - eax 000306a9, ebx 03020800, ecx ffba222b, edx 1f8bfbff - max 13
4: cpu 7 - eax 00000000, ebx 00000000, ecx 00000000, edx 00000000
```

I searched for other code that does AVX2 detection, and it seems it uses 0 for subfunction (2nd parameter).

image

In this instance I am able to get back values:

```
5: cpu 0 - ebx 756e6547, ecx 6c65746e, edx 49656e69 - max 13
5: cpu 1 - eax 000306a9, ebx 03020800, ecx ffba222b, edx 1f8bfbff - max 13
5: cpu 7 - eax 00000000, ebx 00000283, ecx 00000000, edx bc000400
```

**Comment by mtl1979** [comment]
@nmoinvaz Fixed in #528

**Pull request merged**